pkg/lib/ncl/subscriber.go

```diff
@@ -0,0 +1,179 @@
+	}
+}
+
+// WithSubscriberCheckpointer sets the checkpointer for the subscriber
+func WithSubscriberCheckpointer(checkpointer Checkpointer) SubscriberOption {
+	return func(s *subscriber) {
+		s.checkpointer = checkpointer
+	}
+}
+
+// WithSubscriberMessageDeserializer sets the message deserializer for the subscriber
+func WithSubscriberMessageDeserializer(deserializer MessageSerDe) SubscriberOption {
+	return func(s *subscriber) {
+		s.messageDeserializer = deserializer
+	}
+}
+
+// WithSubscriberPayloadRegistry sets the payload registry for the subscriber
+func WithSubscriberPayloadRegistry(registry *PayloadRegistry) SubscriberOption {
+	return func(s *subscriber) {
+		s.payloadRegistry = registry
+	}
+}
+
+// WithSubscriberMessageFilter sets the message filter for the subscriber
+func WithSubscriberMessageFilter(filter MessageFilter) SubscriberOption {
+	return func(s *subscriber) {
+		s.messageFilter = filter
+	}
+}
+
+// defaultSubscriber returns a subscriber with default settings
+func defaultSubscriber(nc *nats.Conn) *subscriber {
+	return &subscriber{
+		nc:                  nc,
+		checkpointer:        &NoopCheckpointer{},
+		messageHandlers:     []MessageHandler{},
+		messageFilter:       NoopMessageFilter{},
+		messageDeserializer: NewEnvelopeSerializer(),
+		subscriptions:       []*nats.Subscription{},
+		mu:                  sync.Mutex{},
+	}
+}
+
+// NewSubscriber creates a new subscriber with the given options
+func NewSubscriber(nc *nats.Conn, opts ...SubscriberOption) (Subscriber, error) {
+	// Start with default subscriber
+	s := defaultSubscriber(nc)
+
+	// Apply all options
+	for _, opt := range opts {
+		opt(s)
+	}
+
+	// Validate the subscriber
+	if err := s.validate(); err != nil {
+		return nil, fmt.Errorf("error validating subscriber: %w", err)
+	}
+
+	return s, nil
+}
+
+// validate checks if the subscriber is properly configured
+func (s *subscriber) validate() error {
+	return errors.Join(
+		validate.NotNil(s.nc, "NATS connection cannot be nil"),
+		validate.NotNil(s.checkpointer, "checkpointer cannot be nil"),
+		validate.NotNil(s.messageDeserializer, "message deserializer cannot be nil"),
+		validate.NotNil(s.payloadRegistry, "payload registry cannot be nil"),
+		validate.IsNotEmpty(s.messageHandlers, "message handlers cannot be empty"),
+		validate.NotNil(s.messageFilter, "message filter cannot be nil"),
+	)
+}
+
+// Subscribe starts consuming messages from the given subjects
+func (s *subscriber) Subscribe(subjects ...string) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	for _, subject := range subjects {
+		sub, err := s.nc.Subscribe(subject, s.processMessage)
+		if err != nil {
+			return err
+		}
+		s.subscriptions = append(s.subscriptions, sub)
+	}
+	return nil
+}
+
+func (s *subscriber) processMessage(m *nats.Msg) {
+	// TODO: implement a better context
+	ctx := context.Background()
+	rMsg := &RawMessage{}
+	if err := s.messageDeserializer.Deserialize(m.Data, rMsg); err != nil {
+		// Handle error
```

Comment:
Don't know if you forgot this, or if it is a future TODO. Would be great to add a TODO in case if it is.

Reply:
yeah meant to deal with this in the future. Added a TODO in https://github.com/bacalhau-project/bacalhau/pull/4257/commits/ccc19e9f3b2d0b16c6a808859b7f1cc7e981c132